akka-stream/src/main/scala/akka/stream/io/SslTlsNew.scala

```diff
@@ -0,0 +1,297 @@
+        if (isAvailable(out)) {
+          byteBuffer.flip()
+          if (byteBuffer.hasRemaining) {
+            push(out, wrapByteString(ByteString.fromByteBuffer(byteBuffer)))
+            byteBuffer.clear()
+            onMoreSpace()
+          } else byteBuffer.clear()
+        } else if (closed) complete(out)
+      }
+
+      final override def onPull(): Unit = flushBytes()
+      final override def onDownstreamFinish(): Unit = onOutputFinished()
+    }
+
+    val engine: SSLEngine = {
+      val e = hostInfo match {
+        case Some((hostname, port)) ⇒ sslContext.createSSLEngine(hostname, port)
+        case None                   ⇒ sslContext.createSSLEngine()
+      }
+
+      e.setUseClientMode(role == Client)
+      e
+    }
+
+    var currentSession = engine.getSession
+
+    /*
+     * So here’s the big picture summary: the SSLEngine is the boss, and it can
+     * be in several states. Depending on this state, we may want to react to
+     * different input and output conditions.
+     *
+     *  - normal bidirectional operation (does both outbound and inbound)
+     *  - outbound close initiated, inbound still open
+     *  - inbound close initiated, outbound still open
+     *  - fully closed
+     *
+     * Upon reaching the last state we obviously just shut down. In addition to
+     * these user-data states, the engine may at any point in time also be
+     * handshaking. This is mostly transparent, but it has an influence on the
+     * outbound direction:
+     *
+     *  - if the local user triggered a re-negotiation, cork all user data until
+     *    that is finished
+     *  - if the outbound direction has been closed, trigger outbound readiness
+     *    based upon HandshakeStatus.NEED_WRAP
+     *
+     * These conditions lead to the introduction of a synthetic TransferState
+     * representing the Engine.
+     */
+
+    var lastHandshakeStatus: HandshakeStatus = _
+    var corkUser = true
+
+    def applySessionParameters(params: NegotiateNewSession): Unit = {
+      import params._
+      enabledCipherSuites foreach (cs ⇒ engine.setEnabledCipherSuites(cs.toArray))
+      enabledProtocols foreach (p ⇒ engine.setEnabledProtocols(p.toArray))
+      clientAuth match {
+        case Some(ClientAuth.None) ⇒ engine.setNeedClientAuth(false)
+        case Some(ClientAuth.Want) ⇒ engine.setWantClientAuth(true)
+        case Some(ClientAuth.Need) ⇒ engine.setNeedClientAuth(true)
+        case None                  ⇒ // do nothing
+      }
+      sslParameters foreach (p ⇒ engine.setSSLParameters(p))
+      engine.beginHandshake()
+      lastHandshakeStatus = engine.getHandshakeStatus
+    }
+
+    def setNewSessionParameters(params: NegotiateNewSession): Unit = {
+      currentSession.invalidate()
+      applySessionParameters(params)
+      corkUser = true
+    }
+
+    // These are Nettys default values
+    // 16665 + 1024 (room for compressed data) + 1024 (for OpenJDK compatibility)
```

Comment:
Perhaps not repeat that calculation multiple times but hoist it into a constant?